### PR TITLE
fix(flow): multi-image paste upload + duplicate handling

### LINKED ISSUE #1802: Duplicate file name error when uploading multiple images from clipboard to Image File Node
\### Self Checks (Required)

- \[x] This is only for bug report, if you would like to ask a question, please head to \[Discussions](https://github.com/giselles-ai/giselle/discussions/categories/q-a).
- \[x] I have searched for existing issues \[search for existing issues](https://github.com/giselles-ai/giselle/issues), including closed ones.

\### Cloud or Self Hosted (Required)

Cloud

\### Version (only for Self Hosted) (Optional)

\### Steps to reproduce (Required)

1.  Go to an 'Image File Node'.
2.  Copy two or more image files from your local file system to the clipboard.
3.  Click on the node and paste the images (e.g., using Ctrl+V or Cmd+V).
4.  See error "duplicate file name".

\### Expected Behavior (Optional)

All selected images should be successfully uploaded to the node without any errors. The system should automatically handle potential file name conflicts, for example, by appending a numerical suffix to duplicate names (e.g., `image.png`, `image-1.png`).

\### Actual Behavior (Optional)

An error message stating "duplicate file name" is shown. The upload process is halted, and none of the images from the clipboard are added to the node.

\### Additional Context (Optional)

When a user attempts to upload two or more images simultaneously from the clipboard to an Image File Node, the process fails and a "duplicate file name" error is displayed. This issue prevents the efficient bulk addition of images via copy-and-paste.

### PR DESCRIPTION
**Pull request by gentamura**
# Fix: Multi-image paste uploads and duplicate handling

Fixes #1802

## Summary
- Ensure a single paste action uploads all images instead of only the last one.
- Update file list using reducer-based `UPDATE_FILE_STATUS` to avoid overwriting with stale node content.
- Treat exact name match (name + extension) as duplicate and skip with clear error: `duplicate file name: <name>`.
- Keep validation consistent between drag-and-drop and paste paths.
- Extract upload logic into `useFileUploads` hook for clarity and reuse.

## Changes
- `packages/giselle/src/react/flow/context.tsx`
  - Replace direct content merging with reducer updates.
  - Improve duplicate detection and error message.
  - Wire in custom hook for upload behavior.
- `packages/giselle/src/react/flow/hooks/use-file-uploads.ts` (new)
  - Handles batching, duplicate checks, step-wise UI updates, and API error handling.
- `packages/giselle/src/react/flow/hooks/index.ts`
  - Export `useFileUploads`.

## Rationale
Previously, multiple uploads in a single paste mutated files based on a captured `node.content`, causing only the last file to persist. Reducer-based updates ensure we always apply changes to the latest state. Duplicate handling prevents redundant files from consuming context with minimal user confusion.

## Validation
- Paste 3 images (e.g., gif/jpg/png with same base name): all 3 upload successfully.
- Paste 3 identical files: shows duplicate error per file and only the first is uploaded.
- Drag-and-drop and file selection follow the same behavior.
- Type-check: `pnpm -F @giselle-sdk/giselle check-types` (passed).

## Risks / Mitigations
- UI updates are now dispatched twice per file (uploading -> final). This improves UX and is batched by store save debounce.
- No change to storage keys (still `fileId`-based), safe for existing data.

## Follow-ups
- Add unit tests around duplicate policies and error handling.
- Consider an opt-in policy flag (error/rename/smart) if product needs evolve.

### CODE REVIEW

File: packages/giselle/src/react/flow/hooks/use-file-uploads.ts [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+						const failedFileData = createFailedFileData(
+							uploadingFileData,
+							message,
+						);
+						fileContents = [
+							...fileContents.filter((f) => f.id !== failedFileData.id),
+							failedFileData,
+						];
+					}
+				}
```
[the remark this answers is not in the document]
Comment (gentamura) [reply packages/giselle/src/react/flow/hooks/use-file-uploads.ts]: Fixed [ae21233](https://github.com/giselles-ai/giselle/pull/1807/commits/ae21233d2ff97a96043b0ca1c7bd00fecb614716)

**Comment by gentamura** [comment]
@toyamarinyon @shige Could you please review this code? 🙏

**Comment by gentamura** [comment]
@shige Since zustand was introduced, the previous modifications weren't taking effect. I've made adjustments to account for zustand and would appreciate your review :pray:

**Comment by gentamura** [comment]
@shige
> When I checked the operation with the latest Preview URL, it seemed to produce the same error as last time. Have you followed the steps at [#1807 (review)](https://github.com/giselles-ai/giselle/pull/1807#pullrequestreview-3199220875) to check the operation?
> 
>  1807-duplicate-error.mp4

Thank you for verifying the functionality.

At the time of last reporting, no errors were displayed during the second paste operation. This update addresses proper display of the duplicate file name error when files share the same name.

image

While currently preventing uploads of files with the same name is by design, would you like to implement a modification to allow uploads of the same-named files? Should we implement a system similar to OS file systems, where files like "image.png" would be renamed to "image copy.png", "image copy 1.png", "image - Copy.png", or "image - Copy (1).png"?

**Comment by shige** [comment]
@gentamura 

> Should we implement a system similar to OS file systems, where files like "image.png" would be renamed to "image copy.png", "image copy 1.png", "image - Copy.png", or "image - Copy (1).png"?

I would like to request this proposal.

---

Below are the ideas proposed by GPT-5.

## Key Idea

When you paste a screenshot (via `Command+V`), browsers typically assign it the default filename `image.png`.
To avoid duplicate filenames, you can intercept the `paste` event, rename the `File` object, and then use it in your upload flow.

---

## Implementation Steps

1. **Listen for the `paste` event** 
Use `document.addEventListener("paste", handler)` to capture pasted images.

2. **Rename the File object** 
Use the `File` constructor: 

```ts 
new File([blob], filename, { type: blob.type }) 
```` 

This allows you to overwrite the default name.

3. **Add a unique suffix** 
Append a value such as: 

* a **timestamp** (`Date.now()`), 
* a **UUID**, 
* or an **incremental counter**.

---

## Example Code (React / Next.js)

```tsx
import { useEffect } from "react";

export default function PasteUploader() { 
useEffect(() => { 
const handlePaste = (e: ClipboardEvent) => { 
const items = e.clipboardData?.items; 
if (!items) return; 

for (const item of items) { 
if (item.type.indexOf("image") === 0) { 
const blob = item.getAsFile(); 
if (!blob) return; 

// Generate a unique filename (using timestamp) 
const timestamp = Date.now(); 
const uniqueFileName = `image-${timestamp}.png`; 

// Rename the File 
const file = new File([blob], uniqueFileName, { type: blob.type }); 

console.log("pasted file:", file); 

// Example: send to your API 
const formData = new FormData(); 
formData.append("file", file); 

fetch("/api/upload", { 
method: "POST", 
body: formData, 
}); 
} 
} 
}; 

document.addEventListener("paste", handlePaste); 
return () => document.removeEventListener("paste", handlePaste); 
}, []); 

return <div>Paste your screenshot with Command+V</div>;
}
````

---

## Filename Uniqueness Options

* **Timestamp:** `image-1694512345678.png`
* **Formatted date:** `image-20250912-103015.png`
* **UUID:** `image-550e8400-e29b-41d4-a716-446655440000.png`
* **Incremental counter:** `image-1.png`, `image-2.png`

**Comment by gentamura** [comment]
@shige The implementation of Filename Uniqueness conflicts in PR #1823, so I plan to address this in a subsequent PR. Therefore, I would appreciate your approval of this PR :pray:

**Comment by gentamura** [comment]
Thank you for reviewers 🚀
